Add support for Get API

Billy99 opened the pull request:
bpfd does not support a Get API. To retrieve a given program, the application must retrieve all programs and filter out the program it needs. This PR adds a Get API to bpfd, which allows an application to retrieve a given program by it's program id. As part of this PR, the Load commands also return the same data returned by the Get for the program that was just loaded.

The API structures were also reworked so that all the Kernel Data is in its own structure and the bpfd based data is in its own structure. These new structs are returned by the List, Get and Load commands.

astoycos commented on proto/bpfd.proto:
```diff
@@ -7,27 +7,71 @@ service Bpfd {
+    AttachInfo attach = 3;
+    map<string, bytes> global_data = 4;
+    optional uint32 map_owner_id = 5;
+    optional string map_pin_path = 6;
```
At least in our internal api map_pin_path isn't optional, and i'm not sure it should be here either.

Billy99 replied:
It's not entered by the user, but returned from bpfd. So needs to be optional.

Billy99 replied:
I'm fine changing it here in the external API to optional, but it is optional in the internal API (command.rs: ProgramData).

Billy99 replied:
Done

astoycos commented on proto/bpfd.proto:
```diff
@@ -75,59 +139,88 @@ message LoadRequest {
-  } 
-  repeated ListResult results = 26;
+  message ListResult { 
+    ProgramInfo info = 1;
```
`ProgramInfo` should be optional here I think since not every program will have bpfd specific informaton

Billy99 replied:
Done

astoycos commented on proto/bpfd.proto:
```diff
@@ -75,59 +139,88 @@ message LoadRequest {
     map<string, string> match_metadata = 3;
 }
 
+/* ListResponse represents a response a request to list loaded and attached
+ * eBPF programs. Based on the corresponding ListRequest programs will include
+ * those loaded by bpfd AND/OR those loaded by other users. The ListResults
+ * will ALWAYS contain a program's kernel info and if loaded by bpfd, will
+ * also contain the state kept by bpfd.
+ */
+
```
```suggestion
/* ListResponse represents a response from listing loaded and attached
 * eBPF programs. Based on the corresponding ListRequest programs will include
 * those loaded by bpfd AND/OR those loaded by other users. The ListResults
 * will ALWAYS contain a program's kernel info and if loaded by bpfd, will
 * also contain the state kept by bpfd.
 */

```

Billy99 replied:
OK

astoycos commented on proto/bpfd.proto:
```diff
@@ -75,59 +139,88 @@ message LoadRequest {
+ * contain the state kept by bpfd. */
+
+message GetResponse { 
+    ProgramInfo info = 1;
```
Optional here as above

Billy99 replied:
Done

astoycos reviewed (changes requested):
Just some comments + nits on the API to start with

Billy99 commented on proto/bpfd.proto:
```diff
@@ -7,27 +7,71 @@ service Bpfd {
+
+message KernelProgramInfo { 
+    uint32 id = 1;
+    /* Same a Name entered by user, but truncated to 16 byte by kernel */
```
OK, tweaked some.
```
    /* Name as seen by the kernel. May be truncated version of user enter */
    /* name since limited to 16 byte by kernel */
```

astoycos commented on bpfd/src/bpf.rs:
```diff
@@ -772,6 +767,26 @@ impl BpfManager {
+            Some(p) => Ok(p.to_owned()),
+            None => loaded_programs()
+                .find_map(|p| {
+                    let prog = p.map_err(BpfdError::BpfProgramError).ok()?;
```
No need to map the error if we're just discarding it :) 
```suggestion
                    let prog = p.ok()?;
```

astoycos reviewed (changes requested):
BPFD Core Review

This just reviews the core bits, I saw you pushed an update while I was reviewing so please ignore any comments that are out of date!

astoycos commented on bpfctl/src/main.rs:
```diff
@@ -663,69 +697,49 @@ impl LoadCommands {
-                location: Some(load_request_common::Location::File(path.clone())),
-                name: name.to_string(),
-                program_type: command.get_prog_type() as u32,
-                global_data: parse_global(global),
-                map_owner_id: *map_owner_id,
+                name: _,
+                metadata: _,
+                global: _,
+                map_owner_id: _,
+                command: _,
+            }) => Ok(Some(BytecodeLocation {
```
```suggestion
            Commands::LoadFromFile(LoadFileArgs {
                path,
                ..
            }) => Ok(Some(BytecodeLocation {
```

astoycos commented on bpfctl/src/main.rs:
```diff
@@ -663,69 +697,49 @@ impl LoadCommands {
-                program_type: command.get_prog_type() as u32,
-                global_data: parse_global(global),
-                map_owner_id: *map_owner_id,
+                name: _,
```
same as above remove all the `_`

astoycos commented on bpfctl/src/main.rs:
```diff
@@ -874,23 +888,70 @@ async fn execute_request_tcp(
+
+            ProgTable::new_get_bpfd(&response.info)?.print();
+            ProgTable::new_get_unsupported(&response.kernel_info)?.print();
+        }
+
+        Commands::LoadFromImage(l) => {
+            let bytecode = match command.get_bytecode_location() {
+                Ok(t) => t,
+                Err(e) => bail!(e),
+            };
+
+            let attach = match command.get_attach_info() {
                 Ok(t) => t,
                 Err(e) => bail!(e),
             };
 
             let request = tonic::Request::new(LoadRequest {
-                common,
-                attach_info,
+                bytecode,
+                name: l.name.to_string(),
+                program_type: l.command.get_prog_type() as u32,
+                attach,
+                metadata: l
+                    .metadata
+                    .clone()
+                    .unwrap_or(vec![])
+                    .iter()
+                    .map(|(k, v)| (k.to_owned(), v.to_owned()))
+                    .collect(),
+                global_data: parse_global(&l.global),
+                uuid: None,
+                map_owner_id: l.map_owner_id,
             });
             let response = client.load(request).await?.into_inner();
-            println!("{}", response.id);
+
+            ProgTable::new_get_bpfd(&response.info)?.print();
+            ProgTable::new_get_unsupported(&response.kernel_info)?.print();
         }
```
Both patterns here are doing the exact same thing can you just do something like this?

```suggestion
        _(l) => {
            let bytecode = match command.get_bytecode_location() {
                Ok(t) => t,
                Err(e) => bail!(e),
            };

            let attach = match command.get_attach_info() {
                Ok(t) => t,
                Err(e) => bail!(e),
            };

            let request = tonic::Request::new(LoadRequest {
                bytecode,
                name: l.name.to_string(),
                program_type: l.command.get_prog_type() as u32,
                attach,
                metadata: l
                    .metadata
                    .clone()
                    .unwrap_or(vec![])
                    .iter()
                    .map(|(k, v)| (k.to_owned(), v.to_owned()))
                    .collect(),
                global_data: parse_global(&l.global),
                uuid: None,
                map_owner_id: l.map_owner_id,
            });
            let response = client.load(request).await?.into_inner();

            ProgTable::new_get_bpfd(&response.info)?.print();
            ProgTable::new_get_unsupported(&response.kernel_info)?.print();
        }
```

Billy99 replied:
I agree on the duplicate code, but stuck on how to do it in rust.

astoycos commented on bpfd-operator/controllers/bpfd-agent/internal/bpfd-core.go:
```diff
@@ -246,21 +223,25 @@ func GetMapsForID(id uint32) (map[string]string, error) {
+			"Verified-Instruction-Count":    fmt.Sprint(kernelInfo.GetVerifiedInsns()),
+		}
 	}
+	return map[string]string{}
```
```suggestion
	return nil
```

Billy99 replied:
Done

astoycos commented on bpfd-operator/controllers/bpfd-agent/common.go:
```diff
@@ -458,7 +458,11 @@ func ProcessMapOwnerParam(ctx context.Context,
-			mapOwnerStatus.mapOwnerId = &prog.Id
+			kernelInfo := prog.GetKernelInfo()
+			if kernelInfo == nil {
+				return nil, fmt.Errorf("failed to process bpfd program for BpfProgram with UUID %s: %v", bpfProgramList.Items[0].GetUID(), err)
```
NIT this and above should be UID not UUID (that's just how it's stated in the k8s spec)

astoycos commented on bpfd-operator/controllers/bpfd-agent/discovered_program.go:
```diff
@@ -100,21 +100,30 @@ func (r *DiscoveredProgramReconciler) Reconcile(ctx context.Context, req ctrl.Re
 			continue
 		}
 
+		bpfProgName := ""
```
nit: move this var back to where it was to avoid the diff

astoycos commented on bpfd-operator/controllers/bpfd-agent/kprobe-program.go:
```diff
@@ -174,30 +174,32 @@ func (r *KprobeProgramReconciler) Reconcile(ctx context.Context, req ctrl.Reques
+
+	return &gobpfd.LoadRequest{
+		Bytecode:    bytecode,
+		Name:        r.currentKprobeProgram.Spec.SectionName,
```
oof we need to change this to `ProgramName` or just `Name` do we have an issue for that?

astoycos commented on bpfd-operator/controllers/bpfd-agent/internal/bpfd-core.go:
```diff
@@ -246,21 +223,25 @@ func GetMapsForID(id uint32) (map[string]string, error) {
 
 // Convert a list result into a set of kernel info annotations
 func Build_kernel_info_annotations(p *gobpfd.ListResponse_ListResult) map[string]string {
```
I didn't see this anywhere, but can you also now make sure to update the BpfPrograms for OUR programs (programs loaded by bpfd) with these annotations as well? 

(Doing this in a separate PR is alright)

Billy99 replied:
#744

astoycos commented on tests/integration-test/src/tests/e2e.rs:
```diff
@@ -341,67 +328,61 @@ fn test_program_execution_with_global_variables() {
+    );
 
-    uuids.push(uuid);
+    uuids.push(prog_id.unwrap());
```
nit: it's probably just ids now but that's alrigt, ultimately all this needs to be deduplicated anyways

Billy99 replied:
Done, was already on my list to do.

astoycos reviewed (changes requested):
Almost there 🥳 !!!!

Billy99 commented on bpfd/src/oci_utils/image_manager.rs:
```diff
@@ -82,18 +82,25 @@ impl BytecodeImage {
-            value.url,
-            value.image_pull_policy,
-            match value.username.as_ref() {
+        let username = if value.username.is_some() {
```
Adding function back. This is what is causing the pull failures. username and password are already Option<String>, but looks like "" is not valid. So this function was converting "" into None.

Billy99 commented on bpfd/src/oci_utils/image_manager.rs:
```diff
@@ -82,18 +82,25 @@ impl BytecodeImage {
 
 impl From<bpfd_api::v1::BytecodeImage> for BytecodeImage {
```
Adding function back.

Billy99 commented:
/hold
I think I found a bug. Let me investigate

Billy99 commented:
Fixed the bug. `bpfctl list --all` was causing bpfd to panic when there was no ProgramInfo (programs not created by bpfd).
```
impl TryFrom<&Program> for V1ProgramInfo {
    type Error = BpfdError;

    fn try_from(program: &Program) -> Result<Self, Self::Error> {
-        let data = program.data().expect("program info not available");
+        let data = program.data()?;
```
And updated the code calling it to map to None when an error was seen (For example: rpc.rs line 129).

astoycos reviewed (approved):
LGTM

Thanks for all the work here @Billy99
